lib/Dialect/FIRRTL/Transforms/LowerDomains.cpp

```diff
@@ -0,0 +1,376 @@
+          paths);
+      PropAssignOp::create(*builder, builder->getUnknownLoc(), subAssociations,
+                           list);
+      // Connect the object to the output port.
+      PropAssignOp::create(*builder, builder->getUnknownLoc(),
+                           body->getArgument(outputPort), object);
+    }
+  }
+
+  // Set new port annotations.
+  op.setPortAnnotationsAttr(ArrayAttr::get(context, portAnnotations));
+
+  LLVM_DEBUG({
+    llvm::dbgs() << "    portMap:\n";
+    for (auto [oldIndex, newIndex] : resultMap)
+      llvm::dbgs() << "      - " << oldIndex << ": " << newIndex << "\n";
+  });
+
+  return success();
+}
+
+LogicalResult LowerModule::lowerInstances(InstanceGraph &instanceGraph) {
+  auto *node = instanceGraph.lookup(cast<igraph::ModuleOpInterface>(*op));
+  for (auto *use : llvm::make_early_inc_range(node->uses())) {
+    auto instanceOp = cast<InstanceOp>(*use->getInstance());
+    LLVM_DEBUG(llvm::dbgs()
+               << "      - " << instanceOp.getInstanceName() << "\n");
+
+    for (auto bit : eraseVector.set_bits()) {
+      auto result = instanceOp.getResult(bit);
+      for (auto *user : llvm::make_early_inc_range(result.getUsers())) {
+        if (auto castOp = dyn_cast<UnsafeDomainCastOp>(user)) {
+          castOp.getResult().replaceAllUsesWith(castOp.getInput());
+          castOp.erase();
+          continue;
+        }
+        user->emitOpError()
+            << "has an unimplemented lowering in the LowerDomains pass.";
+        return failure();
+      }
+    }
+
+    ImplicitLocOpBuilder builder(instanceOp.getLoc(), instanceOp);
+    auto erased = instanceOp.erasePorts(builder, eraseVector);
+    auto inserted = erased.cloneAndInsertPorts(newPorts);
+    for (auto [oldIndex, newIndex] : resultMap) {
+      auto oldPort = erased.getResult(oldIndex);
+      auto newPort = inserted.getResult(newIndex);
+      oldPort.replaceAllUsesWith(newPort);
+    }
+    instanceGraph.replaceInstance(instanceOp, inserted);
+
+    instanceOp.erase();
+    erased.erase();
+  }
+
+  return success();
+}
+
+class LowerCircuit {
+
+public:
+  LowerCircuit(CircuitOp circuit, InstanceGraph &instanceGraph)
+      : circuit(circuit), instanceGraph(instanceGraph) {}
+
+  LogicalResult lowerDomain(DomainOp);
+  LogicalResult lowerCircuit();
+
+private:
+  CircuitOp circuit;
+  InstanceGraph &instanceGraph;
+  DenseMap<Attribute, std::pair<ClassOp, ClassOp>> classes;
+};
+
+LogicalResult LowerCircuit::lowerDomain(DomainOp op) {
+  ImplicitLocOpBuilder builder(op.getLoc(), op);
+  auto *context = op.getContext();
+  auto name = op.getNameAttr();
+  // TODO: Update this once DomainOps have properties.
+  auto classIn = ClassOp::create(builder, name, {});
+  auto classInType = classIn.getInstanceType();
+  auto pathListType =
+      ListType::get(context, cast<PropertyType>(PathType::get(context)));
+  auto classOut =
+      ClassOp::create(builder, StringAttr::get(context, Twine(name) + "_out"),
```

Comment:
If attribute caching, consider caching the field name strings as well.

Reply:
Over-engineered constant pool added in: 30836741f